File: src/main/java/com/freya02/botcommands/internal/prefixed/HelpCommand.java
package com.freya02.botcommands.internal.prefixed;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.prefixed.*;
import com.freya02.botcommands.api.prefixed.annotations.Category;
import com.freya02.botcommands.api.prefixed.annotations.Description;
import com.freya02.botcommands.api.prefixed.annotations.JDATextCommand;
import com.freya02.botcommands.api.prefixed.annotations.TextOption;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.Usability;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.GuildMessageChannel;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.util.*;
import java.util.regex.Pattern;

@Category("Utils")
@Description("Gives help about a command")
public final class HelpCommand extends TextCommand implements IHelpCommand {
	private static final Pattern SPACE_PATTERN = Pattern.compile("\\s+");
	private final BContext context;

	public HelpCommand(BContext context) {
		this.context = context;
	}

	@JDATextCommand(
			name = "help",
			description = "Gives help for all commands"
	)
	public void execute(CommandEvent event) {
		sendGlobalHelp(event);
	}

	@JDATextCommand(
			name = "help",
			description = "Gives help about a command"
	)
	public void execute(BaseCommandEvent event, @TextOption(name = "command path", example = "help") String commandStr) {
		final String[] split = SPACE_PATTERN.split(commandStr);

		if (split.length > 3) {
			event.respond("The command '" + commandStr + "' cannot have more than 3 components").queue();

			return;
		}

		sendCommandHelp(event, CommandPath.of(split));
	}

	@Override
	public void onInvalidCommand(@NotNull BaseCommandEvent event, @NotNull CommandPath executedCommandPath) {
		sendCommandHelp(event, executedCommandPath);
	}

	private void sendGlobalHelp(BaseCommandEvent event) {
		final EmbedBuilder builder = generateGlobalHelp(event.getMember(), event.getGuildChannel());

		final MessageEmbed embed = builder.build();
		event.getAuthor().openPrivateChannel().queue(
				privateChannel -> event.sendWithEmbedFooterIcon(privateChannel, embed, event.failureReporter("Unable to send help message")).queue(
						m -> event.reactSuccess().queue(),
						t -> event.reactError().queue()),
				t -> event.getChannel().sendMessage(context.getDefaultMessages(event.getGuild()).getClosedDMErrorMsg()).queue());

	}

	private void sendCommandHelp(BaseCommandEvent event, CommandPath cmdPath) {
		TextCommandCandidates candidates = event.getContext().findCommands(cmdPath);
		if (candidates == null) {
			event.respond("Command '" + getSpacedPath(cmdPath) + "' does not exist").queue(null, event.failureReporter("Failed to send help"));
			return;
		}

		final Member member = event.getMember();
		final GuildMessageChannel channel = event.getGuildChannel();
		final Usability usability = Usability.of(context, candidates.first(), member, channel, !context.isOwner(member.getIdLong()));
		if (usability.isNotShowable()) {
			event.respond("Command '" + getSpacedPath(cmdPath) + "' does not exist").queue(null, event.failureReporter("Failed to send help"));
			return;
		}

		final EmbedBuilder embed = generateCommandHelp(event, candidates);

		event.respond(embed.build()).queue();
	}

	private EmbedBuilder generateGlobalHelp(Member member, GuildMessageChannel channel) {
		final EmbedBuilder builder = context.getDefaultEmbedSupplier().get();

		builder.setTimestamp(Instant.now());
		builder.setColor(member.getColorRaw());
		builder.setFooter("NSFW commands might not be shown\nRun help in an NSFW channel to see them\n");

		final TreeMap<String, StringJoiner> categoryBuilderMap = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
		final Set<String> insertedCommands = new HashSet<>();
		for (TextCommandCandidates candidates : ((BContextImpl) context).getCommands()) {
			final TextCommandInfo cmd = candidates.first();

			if (insertedCommands.add(cmd.getPath().getName())) { //Prevent duplicates since candidates share the same command path
				if (Usability.of(context, cmd, member, channel, !context.isOwner(member.getIdLong())).isShowable()) {
					categoryBuilderMap
							.computeIfAbsent(Utils.getCategory(cmd), s -> new StringJoiner("\n"))
							.add("**" + cmd.getPath().getName() + "** : " + Utils.getNonBlankDescription(cmd));
				}
			}
		}

		for (Map.Entry<String, StringJoiner> entry : categoryBuilderMap.entrySet()) {
			builder.addField(entry.getKey(), entry.getValue().toString(), false);
		}

		final HelpBuilderConsumer helpBuilderConsumer = context.getHelpBuilderConsumer();
		if (helpBuilderConsumer != null) helpBuilderConsumer.accept(builder, true, null);

		return builder;
	}

	@NotNull
	private EmbedBuilder generateCommandHelp(BaseCommandEvent event, TextCommandCandidates candidates) {
		final EmbedBuilder builder = Utils.generateCommandHelp(candidates, event);
		builder.setTimestamp(Instant.now());
		builder.setColor(event.getMember().getColorRaw());

		final HelpBuilderConsumer helpBuilderConsumer = context.getHelpBuilderConsumer();
		if (helpBuilderConsumer != null) helpBuilderConsumer.accept(builder, false, candidates);

		return builder;
	}

	@NotNull
	private String getSpacedPath(CommandPath cmdPath) {
		return cmdPath.toString().replace('/', ' ');
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/prefixed/HelpCommand.kt
package com.freya02.botcommands.internal.commands.prefixed

import com.freya02.botcommands.api.BContext
import com.freya02.botcommands.api.annotations.CommandMarker
import com.freya02.botcommands.api.commands.annotations.Command
import com.freya02.botcommands.api.commands.prefixed.BaseCommandEvent
import com.freya02.botcommands.api.commands.prefixed.CommandEvent
import com.freya02.botcommands.api.commands.prefixed.IHelpCommand
import com.freya02.botcommands.api.commands.prefixed.TextCommandManager
import com.freya02.botcommands.api.commands.prefixed.annotations.TextDeclaration
import com.freya02.botcommands.api.core.config.BTextConfig
import com.freya02.botcommands.api.core.service.ConditionalServiceChecker
import com.freya02.botcommands.api.core.service.annotations.ConditionalService
import com.freya02.botcommands.api.core.service.annotations.ServiceName
import com.freya02.botcommands.api.core.service.annotations.ServiceType
import com.freya02.botcommands.api.core.service.getInterfacedServices
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.Usability
import com.freya02.botcommands.internal.commands.prefixed.TextUtils.getSpacedPath
import com.freya02.botcommands.internal.simpleNestedName
import com.freya02.botcommands.internal.utils.ReflectionUtils.referenceString
import dev.minn.jda.ktx.coroutines.await
import kotlinx.coroutines.launch
import mu.KotlinLogging
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.exceptions.ErrorResponseException
import net.dv8tion.jda.api.requests.ErrorResponse
import java.time.Instant
import java.util.*

private val logger = KotlinLogging.logger { }
private val spacePattern = Regex("\\s+")

@Command
@ServiceName("builtinHelpCommand")
@ServiceType(IHelpCommand::class)
@ConditionalService(HelpCommand.ExistingHelpChecker::class)
internal class HelpCommand internal constructor(private val context: BContextImpl) : IHelpCommand {
    internal object ExistingHelpChecker : ConditionalServiceChecker {
        override fun checkServiceAvailability(context: BContext, checkedClass: Class<*>): String? {
            // Try to get IHelpCommand interfaced services, except ours
            // If empty, then the user didn't provide one, in which case we can allow
            //Won't take HelpCommand into account
            val helpCommands = context.serviceContainer.getInterfacedServices<IHelpCommand>()
            return when {
                // A user-defined help command was found
                helpCommands.isNotEmpty() -> {
                    logger.debug("Using a custom 'help' text command implementation")
                    "An user supplied IHelpCommand interfaced service is already active (${helpCommands.first()::class.simpleNestedName})"
                }
                // No user-defined help command, try to use ours
                else -> when {
                    context.textConfig.isHelpDisabled -> {
                        logger.debug("Using no 'help' text command implementation")
                        "The help command was disabled in ${BTextConfig::isHelpDisabled.referenceString}"
                    }
                    else -> {
                        logger.debug("Using built-in help command")
                        null
                    }
                }
            }
        }
    }

    @CommandMarker
    suspend fun onTextHelpFallback(event: CommandEvent) {
        sendGlobalHelp(event)
    }

    @CommandMarker
    suspend fun onTextHelpCommand(event: BaseCommandEvent, commandStr: String) {
        val commandInfo = context.textCommandsContext.findTextCommand(spacePattern.split(commandStr))
        if (commandInfo == null) {
            event.respond("Command '$commandStr' does not exist").await()
            return
        }

        sendCommandHelp(event, commandInfo)
    }

    override fun onInvalidCommand(event: BaseCommandEvent, commandInfo: TextCommandInfo) {
        context.coroutineScopesConfig.textCommandsScope.launch {
            sendCommandHelp(event, commandInfo)
        }
    }

    private suspend fun sendGlobalHelp(event: BaseCommandEvent) {
        val builder = generateGlobalHelp(event.member, event.guildChannel)
        val embed = builder.build()

        try {
            event.author
                .openPrivateChannel()
                .flatMap { event.sendWithEmbedFooterIcon(it, embed, event.failureReporter("Unable to send help message")) }
                .await()

            event.reactSuccess().queue()
        } catch (e: ErrorResponseException) {
            when (e.errorResponse) {
                ErrorResponse.CANNOT_SEND_TO_USER -> event.respond(context.getDefaultMessages(event.guild).closedDMErrorMsg).queue()
                else -> throw e
            }
        }
    }

    private suspend fun sendCommandHelp(event: BaseCommandEvent, commandInfo: TextCommandInfo) {
        val member = event.member
        val usability = Usability.of(context, commandInfo, member, event.guildChannel, !context.isOwner(member.idLong))
        if (usability.isNotShowable) {
            event.respond("Command '" + commandInfo.path.getSpacedPath() + "' does not exist").await()
            return
        }

        val embed = generateCommandHelp(event, commandInfo)
        event.respond(embed.build()).queue()
    }

    private fun generateGlobalHelp(member: Member, channel: GuildMessageChannel): EmbedBuilder {
        val builder = context.defaultEmbedSupplier.get()
        builder.setTimestamp(Instant.now())
        builder.setColor(member.colorRaw)
        builder.setFooter("NSFW commands might not be shown\nRun help in an NSFW channel to see them\n")

        val categoryBuilderMap = TreeMap<String, StringJoiner>(String.CASE_INSENSITIVE_ORDER)
        for (cmd in context.textCommandsContext.rootCommands) {
            if (Usability.of(context, cmd, member, channel, !context.isOwner(member.idLong)).isShowable) {
                categoryBuilderMap
                    .computeIfAbsent(cmd.category) { StringJoiner("\n") }
                    .add("**${cmd.name}** : ${cmd.description}")
            }
        }

        for ((key, value) in categoryBuilderMap) {
            builder.addField(key, value.toString(), false)
        }

        context.helpBuilderConsumer?.accept(builder, true, null)

        return builder
    }

    private fun generateCommandHelp(event: BaseCommandEvent, commandInfo: TextCommandInfo): EmbedBuilder {
        val builder = TextUtils.generateCommandHelp(commandInfo, event)
        builder.setTimestamp(Instant.now())
        builder.setColor(event.member.colorRaw)

        context.helpBuilderConsumer?.accept(builder, false, commandInfo)

        return builder
    }

    @TextDeclaration
    internal fun declare(manager: TextCommandManager) {
		manager.textCommand("help") {
            category = "Utils"
            description = "Gives help for a command"

            variation(HelpCommand::onTextHelpCommand) {
                option("commandStr", "command path") {
                    helpExample = "tag"
                }
            }

            variation(HelpCommand::onTextHelpFallback) //fallback
		}
    }
}